server/src/main/java/io/pravega/schemaregistry/service/SchemaRegistryService.java

```diff
@@ -262,30 +266,36 @@ public SchemaRegistryService(SchemaStore store, ScheduledExecutorService executo
         Preconditions.checkArgument(group != null);
         Preconditions.checkArgument(schemaInfo != null);
         log.debug("addSchema called for group {} {}. schema {}", namespace, group, schemaInfo.getType());
-        SchemaInfo schema = normalizeSchemaBinary(schemaInfo);
         // 1. get group policy
         // 2. get checker for serialization format.
         // validate schema against group compatibility policy on schema
         // 3. conditionally update the schema
-        return RETRY.runAsync(() -> store.getGroupEtag(namespace, group)
-                                         .thenCompose(etag ->
-                                                 store.getGroupProperties(namespace, group)
-                                                      .thenCompose(prop -> {
-                                                          return Futures.exceptionallyComposeExpecting(store.getSchemaVersion(namespace, group, schema),
-                                                                  e -> Exceptions.unwrap(e) instanceof StoreExceptions.DataNotFoundException,
-                                                                  () -> { // Schema doesnt exist. Validate and add it
-                                                                      return validateSchema(namespace, group, schema, prop.getCompatibility())
-                                                                              .thenCompose(valid -> {
-                                                                                  if (!valid) {
-                                                                                      throw new IncompatibleSchemaException(String.format("%s is incompatible", schema.getType()));
-                                                                                  }
-                                                                                  return store.addSchema(namespace, group, schema, prop, etag);
-                                                                              });
-                                                                  });
-                                                      })), executor)
+        return RETRY.runAsync(() ->
+                store.getGroupEtag(namespace, group)
+                     .thenCompose(etag ->
+                             store.getGroupProperties(namespace, group)
+                                  .thenCompose(prop -> {
+                                      NormalizedForm normalizedForm = normalizeSchemaBinary(schemaInfo);
+                                      BigInteger fingerprint = normalizedForm.getFingerprint();
+                                      return Futures.exceptionallyComposeExpecting(store.getSchemaVersion(namespace, group, 
```

Comment:
This logic is a bit convoluted actually. I'd think that the common case is the one of adding a schema that doesn't exist yet, but this logic is treating adding a missing schema as the exceptional case. Could we encapsulate this logic in `addSchema` and let it decide whether it exists or not and what to return?

Reply:
we add if it isnt already present. So we first check if its present or not and if its not present we add it. 

esp when schemas are added programatically, then multiple clients could attempt to add the same schema. So by first checking if the schema exists or not, we avoid all the computation for "addschema" upfront.

Reply:
#110